app/models/theme.rb

```diff
@@ -141,33 +185,28 @@ def self.lookup_target(target_id)
     self.targets.invert[target_id]
   end
 
-  def notify_scheme_change(clear_manager_cache = true)
-    Stylesheet::Manager.cache.clear if clear_manager_cache
-    message = refresh_message_for_targets(["desktop", "mobile", "admin"], self)
-    MessageBus.publish('/file-change', message)
-  end
-
-  def notify_theme_change
+  def self.notify_theme_change(theme_ids, with_scheme: false, clear_manager_cache: true)
     Stylesheet::Manager.clear_theme_cache!
+    targets = [:mobile_theme, :desktop_theme]
 
-    themes = [self] + dependant_themes
+    if with_scheme
+      targets.prepend(:desktop, :mobile, :admin)
+      Stylesheet::Manager.cache.clear if clear_manager_cache
+    end
+
+    message = refresh_message_for_targets(targets, theme_ids).flatten
 
-    message = themes.map do |theme|
-      refresh_message_for_targets([:mobile_theme, :desktop_theme], theme)
-    end.compact.flatten
     MessageBus.publish('/file-change', message)
   end
 
-  def refresh_message_for_targets(targets, theme)
+  def notify_theme_change(with_scheme: false)
+    theme_ids = [self.id] + (dependant_themes&.pluck(:id) || [])
```

Comment:
This is creating 3 arrays here. `(dependant_themes&.pluck(:id) || []).unshift(self.id)` will allow us to only create one.